✨ Add fake-ipa (#14)
metal3-io/utility-images: fake-ipa/README.md

> # Fake ironic python agent

Reviewer: ```suggestion
# Fake Ironic Python Agent
```

> Fake ipa is a tool to help test ironic communication with IPA.

Reviewer 2: Let's use some consistent spelling of IPA.

Reviewer: Maybe "productize" it even a bit and consistently call it `FakeIPA` ?

Fake ipa simulate the IPA by:

- Running an API server with the needed real IPA endpoint.
- Send back fake inspection data when requested.

> - lookup the node and save tokens.

Reviewer: ```suggestion
- Lookup the node and save tokens.
```

> - Heartbeating to ironic API with several threads looping over

Reviewer: ```suggestion
- Heartbeating to Ironic API with several threads looping over
```

  a queue of fake agents.

> - Faking the sync/async commands needed by ironic to inspect,

Reviewer: ```suggestion
- Faking the sync/async commands needed by Ironic to inspect,
```

  clean and provision a node.

> ## Run Fake ipa with ironic

Reviewer: Spelling

### Requirements

Machine: `4c / 16gb / 100gb`

> OS: `CentOS9-20220330`

Author: for the host environments

> OS: `CentOS9-20220330`

Reviewer 3: This is a great example of why storing the setup part somewhere else would be beneficial for us. I don't see any reason why we need to use Centos here tbh. Ubuntu would work just fine.

### Test fake ipa

> 1. clone the env scripts `cd` inside `fake-ipa/Run-env` folder

Reviewer: ```suggestion
1. clone the env scripts `cd` inside `fake-ipa/run-env` folder
```
Do not capitalize directory names or file names please.

2. check configs in `config.py`

> 3. run init `./Init-environment.sh`

Reviewer 2: Please, let's not have the third? forth? place where we have virtual machine initialization code. Fake IPA should be integrated as an option to metal3-dev-env.

Author: I am afraid this will add complexity to dev-env and will be hard to understand this is more simple env where we can easily changes and light weight to run

Reviewer 2: I understand, but the crux here is whether you're adding something that only you will be able to use and maintain (then why not personal github) or something that community members will be able to use 1 years from this point while you're on vacation.

We already have a bunch of useful scripts in BMO like `run_local_ironic.sh` that need to be fixed every time someone is trying them.

Reviewer 3: I agree with Dmitry. In the long run this setup should be put somewhere it could be maintained and tested. Dev-env is a good place.
I could live with putting the setup here though, but it needs to be improved. The current state is pretty experimental, and difficult to follow. IMO we could just remove the setup part from this PR, and make some PoC PR where you show how it works. We then can improve from that, and make a "proper" setup.

Reviewer 4: I also agree with Dmitry. I was starting to review the scripts and then I realized that I was having a "deja-vu". This should definitely look better in metal3-dev-env, or at least connected to it in some way to avoid duplication, confusion, or even possible conflicts.

Author: I opened a PR to run fakeIPA on the dev-env https://github.com/metal3-io/metal3-dev-env/pull/1450

> 3. run init `./Init-environment.sh`

Reviewer: Same

4. to just rebuild fake-ipa from the local repo run `./rebuild-fipa.sh`

### Use ironic with fake-ipa

At this step there will be an ironic environment using fake-ipa,
by default there will be two nodes created on ironic you can list them with:
`baremetal node list`

> To manage the a node using the ironicclient:

Reviewer 4: nit: "to manage a node"

> To manage the a node using the ironicclient:

Reviewer: ```suggestion
To manage the a node using the ironic client:
```
or `ironic-client` if you refer to the repo/image.
